src/main/java/htsjdk/samtools/util/ReadEndsIterator.java

```diff
@@ -0,0 +1,207 @@
+     *                     passed in that is not coordinate sorted, it will eventually be coordinated sorted by this class.
+     */
+    public ReadEndsIterator(final SamReader samReader, final IntervalList intervalList) {
+        this(samReader, intervalList, samReader.hasIndex());
+    }
+
+    /**
+     * Prepare to iterate through the given SAM records, skipping non-primary alignments
+     *
+     * @param samReader    must be coordinate sorted
+     * @param intervalList Either the list of desired intervals, or null.  Note that if an intervalList is
+     *                     passed in that is not coordinate sorted, it will eventually be coordinated sorted by this class.
+     * @param useIndex     If true, do indexed lookup to improve performance.  Not relevant if intervalList == null.
+     *                     It is no longer the case the useIndex==true can make performance worse.  It should always perform at least
+     *                     as well as useIndex==false, and generally will be much faster.
+     */
+    public ReadEndsIterator(final SamReader samReader, final IntervalList intervalList, final boolean useIndex) {
+        super(samReader, intervalList, useIndex);
+    }
+
+    /**
+     * Capture the loci covered by the given SAMRecord in the LocusInfos in the accumulator,
+     * creating new LocusInfos as needed. TypedRecordAndOffset object are created only for start
+     * and end of each alignment block of SAMRecord.
+     * If list of intervals is defined, start or/and length of alignment block are shifted to match the interval, to
+     * prevent exceeding the interval.
+     * @param rec SAMRecord to process and add to <code>AbstractLocusInfo</code>
+     */
+    @Override
+    void accumulateSamRecord(SAMRecord rec) {
+        // interpret the CIGAR string and add the base info
+        for (final AlignmentBlock alignmentBlock : rec.getAlignmentBlocks()) {
+            // 0-based offset into the read of the current base
+            int readOffset = alignmentBlock.getReadStart() - 1;
+            // 1-based reference position that the current base aligns to
+            int refPos = alignmentBlock.getReferenceStart();
+
+            // 0-based offset from the aligned position of the first base in the read to the aligned position
+            // of the current base.
+            int refOffset = refPos - rec.getAlignmentStart();
+            int refOffsetEnd = refPos - rec.getAlignmentStart() + alignmentBlock.getLength();
+
+
+            // Ensure there are LocusInfos up to and including this position
+            for (int j = accumulator.size(); j <= refOffsetEnd; ++j) {
+                accumulator.add(createLocusInfo(getReferenceSequence(rec.getReferenceIndex()),
+                        rec.getAlignmentStart() + j));
+            }
+
+            int refOffsetInterval = refOffset;
+            int refOffsetEndInterval = refOffsetEnd;
+            int startShift = 0;
+
```

Comment:
can the general idea be written here in a comment? it would be very helpful for the review process and for later additions. Also, this block of code is very dense and seems the most important part of the PR, could you add comments to the code to explain what is happening?